iface: Add support of `controller` property

cathay4t opened the pull request:
Introducing `BaseInterface.controller` property for attaching port to
existing controller without knowledge of current port list of that
controller.

Example on attach eth1 to br0:

```yaml
---
interfaces:
- name: eth1
  state: up
  controller: br0
```

Unit test cases and integration test case included.

ffmancera commented:
My concerns is that this is not following the declarative design and it is a little bit confusing. Is this property being reported when doing a `show`? If so, how will that work when applying an state that comes from a `show` operation?

What would happen when `controller` property is set and the controller interface state is also present? It will be duplicated?

Another issue I can think on is with this property when an user do a `show` they will get:
```
---
interfaces:
- name: eth1
  state: up
  controller: br0
- name: br0
  state: up
  type: linux-bridge
  bridge:
    options:
      group-forward-mask: 0
      mac-ageing-time: 300
      multicast-snooping: true
      stp:
        enabled: true
        forward-delay: 15
        hello-time: 2
        max-age: 20
        priority: 32768
    port:
      - name: eth1
        stp-hairpin-mode: false
        stp-path-cost: 100
        stp-priority: 32
```

In this situation, if the user wants to deattach the port `eth1`, they will need to remove it from port list in `br0` and also remove the `controller` property in `eth1`. That is confusing in my opinion... In addition, if the user only includes `br0` state without `eth1` will Nmstate fetch the current state and apply `eth1` current `controller` porperty?

I think NMPolicy is the right candidate for this kind of features and as it is now introduced in Nmstate I do see a reason to put this into Nmstate directly..

What do you think? @qinqon @cathay4t

qinqon commented:
> My concerns is that this is not following the declarative design and it is a little bit confusing. Is this property being reported when doing a `show`? If so, how will that work when applying an state that comes from a `show` operation?
> 
> What would happen when `controller` property is set and the controller interface state is also present? It will be duplicated?
> 
> Another issue I can think on is with this property when an user do a `show` they will get:
> 
> ```
> ---
> interfaces:
> - name: eth1
>   state: up
>   controller: br0
> - name: br0
>   state: up
>   type: linux-bridge
>   bridge:
>     options:
>       group-forward-mask: 0
>       mac-ageing-time: 300
>       multicast-snooping: true
>       stp:
>         enabled: true
>         forward-delay: 15
>         hello-time: 2
>         max-age: 20
>         priority: 32768
>     port:
>       - name: eth1
>         stp-hairpin-mode: false
>         stp-path-cost: 100
>         stp-priority: 32
> ```
> 
> In this situation, if the user wants to deattach the port `eth1`, they will need to remove it from port list in `br0` and also remove the `controller` property in `eth1`. That is confusing in my opinion... In addition, if the user only includes `br0` state without `eth1` will Nmstate fetch the current state and apply `eth1` current `controller` porperty?
> 
> I think NMPolicy is the right candidate for this kind of features and as it is now introduced in Nmstate I do see a reason to put this into Nmstate directly..
> 
> What do you think? @qinqon @cathay4t

Adding an interface to an already existing bridge is a common thing, involving nmpolicy for just that is adding more noise, also "controller" field is still declarative, you can re-apply it and the final state will be the same.

So user can remove the port on whatever way they want, changing the bridge or removing controller.

For me the "controller" field makes a lot of sense.

ffmancera commented:
> > My concerns is that this is not following the declarative design and it is a little bit confusing. Is this property being reported when doing a `show`? If so, how will that work when applying an state that comes from a `show` operation?
> > What would happen when `controller` property is set and the controller interface state is also present? It will be duplicated?
> > Another issue I can think on is with this property when an user do a `show` they will get:
> > ```
> > ---
> > interfaces:
> > - name: eth1
> >   state: up
> >   controller: br0
> > - name: br0
> >   state: up
> >   type: linux-bridge
> >   bridge:
> >     options:
> >       group-forward-mask: 0
> >       mac-ageing-time: 300
> >       multicast-snooping: true
> >       stp:
> >         enabled: true
> >         forward-delay: 15
> >         hello-time: 2
> >         max-age: 20
> >         priority: 32768
> >     port:
> >       - name: eth1
> >         stp-hairpin-mode: false
> >         stp-path-cost: 100
> >         stp-priority: 32
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In this situation, if the user wants to deattach the port `eth1`, they will need to remove it from port list in `br0` and also remove the `controller` property in `eth1`. That is confusing in my opinion... In addition, if the user only includes `br0` state without `eth1` will Nmstate fetch the current state and apply `eth1` current `controller` porperty?
> > I think NMPolicy is the right candidate for this kind of features and as it is now introduced in Nmstate I do see a reason to put this into Nmstate directly..
> > What do you think? @qinqon @cathay4t
> 
> Adding an interface to an already existing bridge is a common thing, involving nmpolicy for just that is adding more noise, also "controller" field is still declarative, you can re-apply it and the final state will be the same.
> 
> So user can remove the port on whatever way they want, changing the bridge or removing controller.
> 
> For me the "controller" field makes a lot of sense.

What about this case that I described above? 

```
In this situation, if the user wants to deattach the port eth1, they will need to remove it from port list in br0 and also remove the controller property in eth1. That is confusing in my opinion... In addition, if the user only includes br0 state without eth1 will Nmstate fetch the current state and apply eth1 current controller porperty?
```

qinqon commented:
> > > My concerns is that this is not following the declarative design and it is a little bit confusing. Is this property being reported when doing a `show`? If so, how will that work when applying an state that comes from a `show` operation?
> > > What would happen when `controller` property is set and the controller interface state is also present? It will be duplicated?
> > > Another issue I can think on is with this property when an user do a `show` they will get:
> > > ```
> > > ---
> > > interfaces:
> > > - name: eth1
> > >   state: up
> > >   controller: br0
> > > - name: br0
> > >   state: up
> > >   type: linux-bridge
> > >   bridge:
> > >     options:
> > >       group-forward-mask: 0
> > >       mac-ageing-time: 300
> > >       multicast-snooping: true
> > >       stp:
> > >         enabled: true
> > >         forward-delay: 15
> > >         hello-time: 2
> > >         max-age: 20
> > >         priority: 32768
> > >     port:
> > >       - name: eth1
> > >         stp-hairpin-mode: false
> > >         stp-path-cost: 100
> > >         stp-priority: 32
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > In this situation, if the user wants to deattach the port `eth1`, they will need to remove it from port list in `br0` and also remove the `controller` property in `eth1`. That is confusing in my opinion... In addition, if the user only includes `br0` state without `eth1` will Nmstate fetch the current state and apply `eth1` current `controller` porperty?
> > > I think NMPolicy is the right candidate for this kind of features and as it is now introduced in Nmstate I do see a reason to put this into Nmstate directly..
> > > What do you think? @qinqon @cathay4t
> > 
> > 
> > Adding an interface to an already existing bridge is a common thing, involving nmpolicy for just that is adding more noise, also "controller" field is still declarative, you can re-apply it and the final state will be the same.
> > So user can remove the port on whatever way they want, changing the bridge or removing controller.
> > For me the "controller" field makes a lot of sense.
> 
> What about this case that I described above?
> 
> ```
> In this situation, if the user wants to deattach the port eth1, they will need to remove it from port list in br0 and also remove the controller property in eth1. That is confusing in my opinion... In addition, if the user only includes br0 state without eth1 will Nmstate fetch the current state and apply eth1 current controller porperty?
> ```

We can have the following "mixed" scenarios where a state contains both bridge and the interface 
- Bridge does not have it as a port but interface has "controller" with the bridge
- Bridge has it as a port but interface do not have it as a controller

I would say on both use case interface need to be added as bridge port.

ffmancera commented:
> > > > My concerns is that this is not following the declarative design and it is a little bit confusing. Is this property being reported when doing a `show`? If so, how will that work when applying an state that comes from a `show` operation?
> > > > What would happen when `controller` property is set and the controller interface state is also present? It will be duplicated?
> > > > Another issue I can think on is with this property when an user do a `show` they will get:
> > > > ```
> > > > ---
> > > > interfaces:
> > > > - name: eth1
> > > >   state: up
> > > >   controller: br0
> > > > - name: br0
> > > >   state: up
> > > >   type: linux-bridge
> > > >   bridge:
> > > >     options:
> > > >       group-forward-mask: 0
> > > >       mac-ageing-time: 300
> > > >       multicast-snooping: true
> > > >       stp:
> > > >         enabled: true
> > > >         forward-delay: 15
> > > >         hello-time: 2
> > > >         max-age: 20
> > > >         priority: 32768
> > > >     port:
> > > >       - name: eth1
> > > >         stp-hairpin-mode: false
> > > >         stp-path-cost: 100
> > > >         stp-priority: 32
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > In this situation, if the user wants to deattach the port `eth1`, they will need to remove it from port list in `br0` and also remove the `controller` property in `eth1`. That is confusing in my opinion... In addition, if the user only includes `br0` state without `eth1` will Nmstate fetch the current state and apply `eth1` current `controller` porperty?
> > > > I think NMPolicy is the right candidate for this kind of features and as it is now introduced in Nmstate I do see a reason to put this into Nmstate directly..
> > > > What do you think? @qinqon @cathay4t
> > > 
> > > 
> > > Adding an interface to an already existing bridge is a common thing, involving nmpolicy for just that is adding more noise, also "controller" field is still declarative, you can re-apply it and the final state will be the same.
> > > So user can remove the port on whatever way they want, changing the bridge or removing controller.
> > > For me the "controller" field makes a lot of sense.
> > 
> > 
> > What about this case that I described above?
> > ```
> > In this situation, if the user wants to deattach the port eth1, they will need to remove it from port list in br0 and also remove the controller property in eth1. That is confusing in my opinion... In addition, if the user only includes br0 state without eth1 will Nmstate fetch the current state and apply eth1 current controller porperty?
> > ```
> 
> We can have the following "mixed" scenarios where a state contains both bridge and the interface
> 
>     * Bridge does not have it as a port but interface has "controller" with the bridge
> 

This is a problem IMO. This would break our API as we always let them know that if the bridge interface is in desired state with the port property set, these are the ports that are going to end up on in the kernel config. With this feature, the users will need to remove it from both places and if they miss one, it will set the interface as a port.

IMO, it would be better if the controller property is not reported when doing a `show` and the `controller` property is something that the user use as a shortcut for adding a port to a controller but it won't be present on the current state. Maybe I am missing something.. but I still think this have the same problems as the `copy-ip-address` we discussed in the past.

>     * Bridge has it as a port but interface do not have it as a controller
> 
> 
> I would say on both use case interface need to be added as bridge port.

cathay4t commented:
This is unlike `copy-ip-address`, the desire state in comment 0 can be applied idempotent.

For the conflicting `controller` and port list, user will get failure when miss-match.

For extra work when detaching ports, we will hide `controller` property when show or `retrieve()`.

cathay4t commented:
New patch uploaded with changes:
 * Checking on these invalid use cases:
    * eth1 has controller: br1, but br1 has no eth1 in port list
    * eth2 has controller: br1, but br2 has eth2 in port list
    * eth1 has controller: `Some("")` (means detach), but br1 has eth1 in port list
 * Hide `controller` from querying and showing.
 * Rust in-code document updated for above.
 * Unit test case for asserting above invalid use cases.

cathay4t committed:
iface: Add support of `controller` property

Introducing `BaseInterface.controller` property for attaching port to
existing controller without knowledge of current port list of that
controller.

Example on attach eth1 to br0:

```yaml
---
interfaces:
- name: eth1
  state: up
  controller: br0
```

The `controller` property is only valid for applying, always been set to
None(hidden during serializing) on show/`retrieve()`.

Unit test cases and integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera commented:
> This is unlike `copy-ip-address`, the desire state in comment 0 can be applied idempotent.
> 
> For the conflicting `controller` and port list, user will get failure when miss-match.
> 
> For extra work when detaching ports, we will hide `controller` property when show or `retrieve()`.

After these changes it looks good to me. It is a little bit weird that one interface state is modifying a different interface state but considering there is a `controller` -> `port` relationship, it is fine I think. Will review, thanks!
